### ISSUE TITLE
Why predict same test set, but given different predict_prob?

### ISSUE THREAD

**Issue by BigHeartC** [ChenglongChen/tensorflow-DeepFM#10]
After train a DeepFM model, I call predict() many times with same test set, but given predict_prob is slightly different.
such as:

> array([ 0.3498897 ,  0.30687785,  0.34534037], dtype=float32)  #first time call predict()

> array([ 0.3498849 ,  0.30688545,  0.34534937], dtype=float32) #second time call predict()

**Comment by ChenglongChen** [comment]
Are you using GPU? Can you check whether this happen again switching to CPU?

**Comment by BigHeartC** [comment]
It's happen when I set "batch_norm": 1, otherwise it will not happen.

**Comment by sudongxiang** [comment]
i have  the  same  isue ； however, I  find  that  batch_norm  is  not  working  for distributed  training;  so, just  disable  batch_norm;

**Comment by hugstone** [comment]
> After train a DeepFM model, I call predict() many times with same test set, but given predict_prob is slightly different.
> such as:
> 
> > array([ 0.3498897 ,  0.30687785,  0.34534037], dtype=float32)  #first time call predict()
> 
> > array([ 0.3498849 ,  0.30688545,  0.34534937], dtype=float32) #second time call predict()

in predict online，i have the same issue.  the problem is in "batch_norm_layer"， tf.cond is not a lazy operator，when predict，bn_train is always excuted, so the moving_mean an moving_variance is always updated online. the solution is "z = tf.cond(train_phase, lambda: self.get_bn_train(x, scope_bn), lambda: self.get_bn_inference(x, scope_bn))"
